combined-docker/gdp/apps/gdp-create.8
.Dd August 2, 2015
.Dt GDP-CREATE 8
.Os Swarm-GDP
.Sh NAME
.Nm gdp-create
.Nd create new GDP log
.Sh SYNOPSIS
.Nm gdp-create
.Op Fl C Ar creator-name
.Op Fl D Ar debug-spec
.Op Fl e Ar key-crypto
.Op Fl G Ar router-ip
.Op Fl h Ar hash-alg
.Op Fl k Ar key-type
.Op Fl K Ar owner-key-location
.Op Fl b Ar key-bits
.Op Fl c Ar curve-name
.Op Fl q
.Op Fl s Ar logd-gdp-addr
.Op Fl S
.Op Fl w
.Op Fl W Ar writer-key-location
.Op Ar mdid=metadata ...
.Op Ar external-name
.Sh DESCRIPTION
.Nm gdp-create
creates a new GDP object,
a.k.a.
.Ql "Data Capsule"
(sometimes called a
.Ql "log"
mostly for historic reasons)
on a given log server.
In its default mode it also creates a public/secret
.Ql "owner"
keypair associated with this object,
with the public key stored in the object metadata.
Additional (text-based) metadata can be added
(arguably this should allow binary metadata as well).
Optionally the log can be given a human-understandable name;
if not specified a seemingly random name is made up.
.
.Pp
The human-oriented name has certain requirements:
.Bl -bullet
.
.It
It may not look like the base64-encoded internal name of any log.
.
.It
It may not already be listed in the Human-Oriented to GDPname Directory (HONGD).
.
.It
It may not have an equal sign in it.
.El
.
.Pp
Also, if the human-oriented name has no dots but the
.Ev GDP_NAME_ROOT
environment variable is set,
the two will be appended in the same way as described in
.Xr gdp 7 .
.
.Ss "Metadata"
User-specified metadata may be added when a log is created.
The metadata name is stored as a 32-bit number and is generally represented
by four characters.
.Pp
In addition to user metadata,
.Nm gdp-create
adds metadata for the log creation time,
the public half of the signing key in DER format,
and the human-readable name.
.Pp
Metadata is immutable; there is no way to add, delete, or change metadata
after the log is created.
.
.Ss "Key Management"
GDP objects may have two secret keys associated with them:
the
.Ql owner
key allows administrative control over the object and the
.Ql writer
key allows write access to the object.
If the writer key is not specifies it defaults to the same as the owner key.
.
.Ss "Warning"
This implementation will change in the future,
and the command will probably be subsumed into a swarm service.
.Sh OPTIONS
.Bl -tag
.
.It Fl C
Sets the name of the creator in the log metadata.
If this is not specified it will be derived from system data.
It should be in the form of an email address.
.
.It Fl D Ar debug-spec
Turns on debugging flags using the libep-style format.
Useful only with the code in hand.
.
.It Fl e Ar key-crypto
Specifies symmetric encryption algorithm to use for storing the secret key.
If
.Ar key-crypto
is
.Li none
then the key is stored unencrypted on disk.
Defaults to
.Va swarm.gdp.crypto.keyenc.alg .
.
.It Fl G Ar router-ip
Contact the given IP address (which may be a DNS name)
to get a connection to the GDP routing layer.
This will be replaced by automatic discovery
(for example, using Zeroconf)
soon.
The address may have an optional port number after a colon.
If this is not specified,
the names listed in the
.Va swarm.gdp.routers
runtime administrative parameter
are searched.
.
.It Fl h Ar hash-alg
Selects the hash (message digest) algorithm used when
generating and checking signatures.
When data is appended to the log,
the log name, log metadata, record number, and record data
is hashed using this algorithm and then that hash is signed.
.
.It Fl k Ar key-type
Selects the type of the creator key to be created for this log.
It can be one of
.Li rsa ,
.Li dsa ,
or
.Li ec .
As a special case, the type
.Li none
suppresses creation of the keypair.
The creator key has both administrative and write access to the log.
.
.It Fl K Ar owner-key-location
This option has two purposes.
First, if you want to use an existing keypair
(formatted as a PEM file)
as the owner key
you can use the
.Fl K
flag to name an existing PEM file.
If you want
.Nm
to create a keypair for you,
.Fl K
can name a directory into which the secret key will be written;
the file will be named with the internal name of the log with
.Va \&-owner.pem
appended.
If
.Fl K
is not specified, the value of the
.Va swarm.gdp.crypto.key.dir
administrative parameter will be used;
it defaults to
.Pa ./KEYS ,
or to the current directory if the
.Pa KEYS
directory does not exist.
.
.It Fl b Ar key-bits
If the key type is RSA or DSA,
this parameter specifies the number of bits in the key.
If this is omitted and the key type is RSA,
this defaults to the value of
.Va swarm.gdp.crypto.rsa.keylen .
DSA keys use
.Va swarm.gdp.crypto.dsa.keylen .
Additionally, RSA keys use the
.Va swarm.gdp.crypto.rsa.keyexp
parameter to set the exponent.
This parameter is ignored when creating EC key pairs.
.
.It Fl c Ar curve-name
Sets the name of the curve to use for Elliptic Cryptography keys.
This overrides the
.Va swarm.gdp.crypto.ec.curve
parameter.
Ignored when creating RSA or DSA key pairs.
.
.It Fl q
If set,
.Nm
does not show errors or informational message.
In particular, using
.Fl q
is a way to ensure that a log exists without give an error if it already exists.
.
.It Fl s Ar logd-gdp-addr
The name of the log daemon on which to create the physical instantiation
of the log.
If not specified, the value of the
.Va swarm.gdp.creation-service.name
administrative will be used.
If that is not set,
it will default to one of the servers at Berkeley.
If you want a specific server at your site,
either use the
.Fl s
flag or set that administrative parameter.
.
.It Fl S
Skips the test to see if the log already exists before creating it.
Used for debugging.
.It Fl w
Create a separate writer key.
The writer key does not convey administrative access to the log.
The owner key conveys write access to the log even if a separate
writer key is created.
.
.It Fl W Ar writer-key-location
Similar to
.Fl K ,
but for the writer key.
.El
.Sh EXIT STATUS
.Bl -tag
.It Li EX_OK No (0)
Success
.It Li EX_CANTCREAT No (73)
The log already exists.
.It Li EX_DATAERR No (65)
A command-line key file could not be parsed.
.It Li EX_IOERR No (74)
A secret key file could not be created.
.It Li EX_NOHOST No (68)
The log server name was not valid.
.It Li EX_NOINPUT No (67)
A command-line key file could not be opened.
.It Li EX_SOFTWARE No (70)
A successfully created public key could not be converted to external (DER) form
for storage in new log metadata,
or severe internal error.
.It Li EX_USAGE No (64)
Command line parameters are incorrect.
.It Li EX_UNAVAILABLE No (69)
A specified key length was insecure.
It was impossible to create a new key.
Other unspecified error.
.El
.Sh ADMINISTRATIVE PARAMETERS
.Bl -tag
.It swarm.gdp.crypto.dsa.keylen
The DSA key length.
Defaults to 2048.
Overridden by
.Fl b .
.It swarm.gdp.crypto.ec.curve
The EC curve.
Defaults to
.Li sect283r1
(also known as
.Li B-283 ) .
Overridden by
.Fl c .
.It swarm.gdp.crypto.hash.alg
The hash algorithm.
Defaults to
.Li sha256 .
Overridden by
.Fl h .
.It swarm.gdp.crypto.key.dir
The directory in which to store secret keys.
Defaults to
.Pa KEYS
(in the current working directory).
If that subdirectory does not exist,
the keys are written to the current directory.
Overridden by
.Fl K .
.It swarm.gdp.crypto.keyenc.alg
The secret key (symmetric) encryption algorithm.
Defaults to
.Li aes192 .
Overridden by
.Fl e .
.It swarm.gdp.crypto.rsa.keyexp
The exponent to be used in the RSA algorithm.
Defaults to 3.
.It swarm.gdp.crypto.rsa.keylen
The key length for the RSA algorithm.
Defaults to 2048.
Overridden by
.Fl b .
.It swarm.gdp.crypto.sign.alg
The signing algorithm.
Defaults to
.Li ec .
Overridden by
.Fl k .
.
.It swarm.gdp.creation-service.name
The GDPname of the creation service.
Overridden by
.Fl s .
.El
.Sh ENVIRONMENT
.Bl -tag
.
.It GDP_NAME_ROOT
If set, this is prepended to any human-oriented names
that do not already have dots in them.
This is an easy way to make your log names unique.
See
.Xr gdp 7
for details.
.\".Sh FILES
.Sh SEE ALSO
.Xr gdp 7 ,
.Xr gdplogd 8
.Sh EXAMPLES
.Bl -bullet
.It
To create a GDP log on a default server:
.Dl newlog=edu.berkeley.cs.eric.example
.Dl gdp-create $newlog
The secret key will be written into a a file named
.Pa KEYS/<something>.pem ,
where
.Va <something>
is the base-64-encoded internal name of the log, e.g.,
.Li 1KZy5jy1QpghTe8QBmDQGqdz3a_9tVP3qp6uxlOeJdk.pem .
The password for encrypting this key will be read from the standard input.
.It
To create a GDP log named
.Li $newlog
on log server named edu.berkeley.eecs.gdp-01 using defaults:
.Dl logd=edu.berkeley.eecs.gdp-01.gdplogd
.Dl gdp-create -s $logd $newlog
.It
To create a log on the server named
.Li $logd
encrypted with a 1024-bit RSA key,
leaving the unencrypted key in a file named
.Pa mykey.pem :
.Dl gdp-create -k RSA -b 1024 -e none -K mykey.pem $newlog
.It
To create a log with user-specified metadata:
.Dl gdp-create Qo "MYMD=My special metadata" Qc $newlog
.It
To create a log without a human-friendly name using sha-224
as the hash (message digest) algorithm:
.Dl gdp-create -h sha224
.Nm gdp-create
will print the base-64-encoded name of the new log.
You should be careful to record that,
since it is your only way to access the log.
.El
.
.Sh BUGS
There should be some way to provide the key password
other than reading it from the standard input.
.Pp
It should not be necessary to name the server on which the log will be created.
This command will be replaced by a smart service
that chooses placement based on locality and capacity.


combined-docker/gdp/apps/gdp-create.c
/* vim: set ai sw=4 sts=4 ts=4 : */

/*
**  LOG-CREATE --- create a new log
**
**		This is a temporary interface.  Ultimately we need a log
**		creation service that does reasonable log placement rather
**		than having to name a specific log server.
**
**	----- BEGIN LICENSE BLOCK -----
**	Applications for the Global Data Plane
**	From the Ubiquitous Swarm Lab, 490 Cory Hall, U.C. Berkeley.
**
**	Copyright (c) 2015-2019, Regents of the University of California.
**	All rights reserved.
**
**	Permission is hereby granted, without written agreement and without
**	license or royalty fees, to use, copy, modify, and distribute this
**	software and its documentation for any purpose, provided that the above
**	copyright notice and the following two paragraphs appear in all copies
**	of this software.
**
**	IN NO EVENT SHALL REGENTS BE LIABLE TO ANY PARTY FOR DIRECT, INDIRECT,
**	SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES, INCLUDING LOST
**	PROFITS, ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION,
**	EVEN IF REGENTS HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
**
**	REGENTS SPECIFICALLY DISCLAIMS ANY WARRANTIES, INCLUDING, BUT NOT
**	LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
**	FOR A PARTICULAR PURPOSE. THE SOFTWARE AND ACCOMPANYING DOCUMENTATION,
**	IF ANY, PROVIDED HEREUNDER IS PROVIDED "AS IS". REGENTS HAS NO
**	OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS,
**	OR MODIFICATIONS.
**	----- END LICENSE BLOCK -----
*/

#include <ep/ep.h>
#include <ep/ep_app.h>
//#include <ep/ep_crypto.h>
#include <ep/ep_dbg.h>
//#include <ep/ep_string.h>
#include <gdp/gdp.h>

#include <errno.h>
#include <string.h>
#include <sysexits.h>
#include <sys/stat.h>


//static EP_DBG	Dbg = EP_DBG_INIT("gdp-create", "Create new log");


// command line options
#define OPTIONS		"b:c:C:D:e:G:h:k:K:qs:SwW:"

void
usage(void)
{
	fprintf(stderr, "Usage: %s [-C creator] [-D dbgspec] [-e key_enc_alg]\n"
			"\t[-G gdpd_addr] [-q] [-s logd_name] [-S]\n"
			"\t[-h] [-k keytype] [-K owner_keyfile] [-w] [-W writer_keyfile]\n"
			"\t[-b keybits] [-c curve] [<mdid>=<metadata>...] [log_name]\n"
			"    -C  set name of log creator (owner; for metadata)\n"
			"    -D  set debugging flags\n"
			"    -e  set secret key encryption algorithm\n"
			"    -G  IP host to contact for GDP router\n"
			"    -q  don't give error if log already exists\n"
			"    -s  set creation service name\n"
			"    -S  skip the test to see if log already exists (for debugging)\n"
			"    -h  set the hash (message digest) algorithm (defaults to sha256)\n"
			"    -k  type of key; valid key types are \"rsa\", \"dsa\", and \"ec\"\n"
			"\t(defaults to ec); \"none\" turns off key generation\n"
			"    -K  use indicated key/place to write secret owner key\n"
			"    -w  create separate writer key\n"
			"    -W  use indicated key/place to write secret writer key\n"
			"\tIf -K or -W specify a directory, a .pem file is written there\n"
			"\twith the name of the GCL (defaults to \"KEYS\" or \".\")\n"
			"    -b  set size of key in bits (RSA and DSA only)\n"
			"    -c  set curve name (EC only)\n"
			"    creator is the id of the creator, formatted as an email address\n"
			"    logd_name is the name of the log server to host this log\n"
			"    metadata ids are (by convention) four letters or digits\n"
			"    log_name is the name of the log to create\n",
			ep_app_getprogname());
	exit(EX_USAGE);
}


int
main(int argc, char **argv)
{
	const char *gobxname = NULL;	// external name of GCL
	const char *logdxname = NULL;	// external name of log daemon
	const char *cname = NULL;		// creator/owner name (for metadata)
	gdp_gin_t *gin = NULL;
	gdp_create_info_t *gci = NULL;
	int opt;
	EP_STAT estat;
	char *gdpd_addr = NULL;
	bool show_usage = false;
	bool quiet = false;
	bool skip_existence_test = false;
	const char *dig_alg_name = "def";
	const char *key_alg_name = "def";
	int key_bits = 0;
	const char *key_curve = NULL;
	const char *owner_keyfile = NULL;
	const char *writer_keyfile = NULL;
	const char *key_enc_alg_name = "def";
	bool separate_writer_key = false;
	const char *phase = NULL;

	// quick pass so debugging is on for initialization
	while ((opt = getopt(argc, argv, OPTIONS)) > 0)
	{
		if (opt == 'D')
			ep_dbg_set(optarg);
	}
	optind = 1;
#if EP_OSCF_NEED_OPTRESET
	optreset = 1;
#endif

	// preinit library (must be early due to crypto code in arg processing)
	ep_crypto_init(0);

	// collect command-line arguments
	while ((opt = getopt(argc, argv, OPTIONS)) > 0)
	{
		switch (opt)
		{
		 case 'b':
			key_bits = atoi(optarg);
			break;

		 case 'c':
			key_curve = optarg;
			break;

		 case 'C':
			cname = optarg;
			break;

		 case 'D':
			// already done
			break;

		 case 'e':
			if (ep_crypto_keyenc_byname(optarg) < 0)
			{
				ep_app_error("unknown key encryption algorithm %s", optarg);
				show_usage = true;
			}
			key_enc_alg_name = optarg;
			break;

		 case 'G':
			gdpd_addr = optarg;
			break;

		 case 'h':
			if (ep_crypto_md_alg_byname(optarg) < 0)
			{
				ep_app_error("unknown digest hash algorithm %s", optarg);
				show_usage = true;
			}
			dig_alg_name = optarg;
			break;

		 case 'k':
			if (ep_crypto_keytype_byname(optarg) == EP_CRYPTO_KEYTYPE_UNKNOWN)
			{
				ep_app_error("unknown key type %s", optarg);
				show_usage = true;
			}
			key_alg_name = optarg;
			break;

		 case 'K':
			owner_keyfile = optarg;
			break;

		 case 'q':
			quiet = true;
			break;

		 case 's':
			logdxname = optarg;
			break;

		 case 'S':
			skip_existence_test = true;
			break;

		 case 'w':
			separate_writer_key = true;
			break;

		 case 'W':
			writer_keyfile = optarg;
			separate_writer_key = true;
			break;

		 default:
			show_usage = true;
			break;
		}
	}
	argc -= optind;
	argv += optind;

	if (show_usage || argc < 0)
		usage();

	// initialize the GDP library
	phase = "initializing GDP";
	estat = gdp_init(gdpd_addr);
	if (!EP_STAT_ISOK(estat))
		goto fail0;

	// allow thread to settle to avoid interspersed debug output
	ep_time_nanosleep(INT64_C(100000000));

	gci = gdp_create_info_new();

	// collect any metadata and external name
	while (argc > 0)
	{
		char *p;
		if ((p = strchr(argv[0], '=')) != NULL)
		{
			gdp_md_id_t mdid = 0;
			int i;

			p++;
			for (i = 0; i < 4; i++)
			{
				if (argv[0][i] == '=')
					break;
				mdid = (mdid << 8) | (unsigned) argv[0][i];
			}
			if (argv[0][i] != '=')
				ep_app_warn("metadata id %.*s truncated to 4 characters",
						(int) (p - argv[0] - 1), argv[0]);

			gdp_create_info_add_metadata(gci, mdid, strlen(p), p);
		}
		else if (gobxname == NULL)
		{
			gobxname = *argv;
		}
		else
		{
			usage();
		}

		argc--;
		argv++;
	}

	// name is optional ; if omitted one will be created
	if (argc-- > 0)
		gobxname = *argv++;
	if (gobxname != NULL && gobxname[0] == '0')
		gobxname = NULL;

	if (show_usage || argc > 0)
		usage();

	if (gobxname != NULL)
	{
		// check validity of the name
		gdp_name_t gobiname;
		if (EP_STAT_ISOK(gdp_internal_name(gobxname, gobiname)))
		{
			// this is a valid base64-encooded log name: not appropriate
			estat = GDP_STAT_GDP_NAME_INVALID;
			if (!quiet)
			{
				ep_app_error("Cannot choose internal GDPname (%s)",
						gobxname);
				exit(EX_CANTCREAT);
			}
		}
		if (!skip_existence_test)
		{
			/*
			**  If the external name is already known in the
			**  Human-Oriented Name to GDPname directory, OR if the
			**  hashed (old-style) internal name can be found,
			**  then there is a conflict.
			*/

			gin = NULL;
			estat = gdp_name_parse(gobxname, gobiname, NULL);
			if (EP_STAT_ISWARN(estat))
			{
				// iname used is the SHA256 of the xname
				estat = gdp_gin_open(gobiname, GDP_MODE_RO, NULL, &gin);
			}
			if (EP_STAT_ISOK(estat))
			{
				// oops, we shouldn't be able to parse or open it
				if (gin != NULL)
					(void) gdp_gin_close(gin);
				if (!quiet)
					ep_app_error("Cannot create %s: already exists", gobxname);
				exit(EX_CANTCREAT);
			}
		}
	}

	/**************************************************************
	**  Set up other automatic metadata
	*/

	if (cname != NULL)
	{
		phase = "setting creator";
		estat = gdp_create_info_set_creator(gci, cname, NULL);
		EP_STAT_CHECK(estat, goto fail1);
	}

	if (logdxname != NULL)
	{
		phase = "setting creation service";
		estat = gdp_create_info_set_creation_service(gci, logdxname);
		// OK if iname is hash of xname, which presents as a warning
		if (EP_STAT_ISFAIL(estat))
			goto fail1;
	}

	if (owner_keyfile != NULL)
	{
		// could be an existing file or a directory
		phase = "setting owner key";
		struct stat st;
		int istat = stat(owner_keyfile, &st);

		if (istat < 0)
		{
			// non-existent file: error
			ep_app_error("%s must exist", owner_keyfile);
			estat = ep_stat_from_errno(ENOENT);
		}
		else if ((st.st_mode & S_IFMT) == S_IFDIR)
		{
			// arrange for key to be written to this directory
			phase = "creating owner key";
			ep_adm_setparam("swarm.gdp.crypto.key.dir", owner_keyfile);
			estat = gdp_create_info_new_owner_key(gci, dig_alg_name,
							key_alg_name, key_bits, key_curve,
							key_enc_alg_name);
		}
		else
		{
			// existing key file --- use it instead of creating new key
			EP_CRYPTO_KEY *key = NULL;
			key = ep_crypto_key_read_file(owner_keyfile,
								EP_CRYPTO_KEYFORM_UNKNOWN, EP_CRYPTO_F_SECRET);
			estat = gdp_create_info_set_owner_key(gci, key, dig_alg_name);
		}
	}
	else
	{
		phase = "creating owner key";
		estat = gdp_create_info_new_owner_key(gci, dig_alg_name, key_alg_name,
						key_bits, key_curve, key_enc_alg_name);
	}
	EP_STAT_CHECK(estat, goto fail1);

	if (writer_keyfile != NULL)
	{
		phase = "setting writer key";
		EP_CRYPTO_KEY *key;
		key = ep_crypto_key_read_file(writer_keyfile, 0, EP_CRYPTO_F_SECRET);
		estat = gdp_create_info_set_writer_key(gci, key, dig_alg_name);
	}
	else if (separate_writer_key)
	{
		phase = "creating writer key";
		estat = gdp_create_info_new_writer_key(gci, dig_alg_name, key_alg_name,
						key_bits, key_curve, key_enc_alg_name);
	}
	EP_STAT_CHECK(estat, goto fail1);

	// if the user specified an unqualified name and we have a root name,
	// use the extended name.
	const char *root = gdp_name_root_get();
	char xnamebuf[GDP_HUMAN_NAME_MAX + 1];
	if (gobxname != NULL && strchr(gobxname, '.') == NULL && root != NULL)
	{
		snprintf(xnamebuf, sizeof xnamebuf, "%s.%s", root, gobxname);
		gobxname = xnamebuf;
	}

	/*
	**  Hello sailor, this is where the actual creation happens
	*/

	phase = "creating GDP object";
	estat = gdp_gin_create(gci, gobxname, &gin);
	EP_STAT_CHECK(estat, goto fail1);

	// just for a lark, let the user know the (internal) name
	if (!quiet)
	{
		gdp_pname_t pname;

		// this output gets parsed by gdp-rest --- don't change it!
		printf("Created new GCL %s\n",
				gdp_printable_name(*gdp_gin_getname(gin), pname));
	}

	gdp_gin_close(gin);

fail1:
	// free creation information
	gdp_create_info_free(&gci);

fail0:
	if (!EP_STAT_ISOK(estat))
	{
		char ebuf[100];
		ep_app_error("Creation failed while %s: %s",
				phase, ep_stat_tostr(estat, ebuf, sizeof ebuf));
	}

	int exitstat;
	if (EP_STAT_ISOK(estat))
		exitstat = EX_OK;
	else if (EP_STAT_IS_SAME(estat, GDP_STAT_NAK_NOROUTE))
		exitstat = EX_NOHOST;
	else if (EP_STAT_IS_SAME(estat, GDP_STAT_NAK_CONFLICT))
		exitstat = EX_CANTCREAT;
	else if (EP_STAT_ISABORT(estat))
		exitstat = EX_SOFTWARE;
	else
		exitstat = EX_UNAVAILABLE;

	// OK status can have values; hide that from the user
	if (EP_STAT_ISOK(estat))
		estat = EP_STAT_OK;

	if (!EP_STAT_ISOK(estat))
		ep_app_message(estat, "exiting with status");
	else if (!quiet)
		fprintf(stderr, "Exiting with status OK\n");

	return exitstat;
}
